packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -159,46 +172,88 @@ abstract class GlobalKey<T extends State<StatefulWidget>> extends Key {
-          ),
-        ]);
+      assert(newChild != null);
+      _debugReservations[parent] ??= <Element, GlobalKey>{};
+      // Removes the old children reservation if exist.
+      if (oldChild != null) {
+        _debugRemoveReservationFor(parent, oldChild);
       }
-      _debugReservations[this] = parent;
+      _debugReservations[parent][newChild] = this;
+      return true;
+    }());
+  }
+
+  static void _debugVerifyGlobalKeyReservation() {
+    assert(() {
+      final Map<GlobalKey, Element> keyToParent = <GlobalKey, Element>{};
+      _debugReservations.forEach((Element parent, Map<Element, GlobalKey> chidToKey) {
+        // We ignore parent that are detached.
+        if (parent.renderObject?.attached == false)
+          return;
+        chidToKey.forEach((Element child, GlobalKey key) {
+          // If parent = null, the node is deactivated by its parent and is
+          // not re-attached to other part of the tree. We should ignore this
+          // node.
+          if (child._parent == null)
+            return;
+          // It is possible the same key registers to the same parent twice
+          // with different children. That is illegal, but it is not in the
+          // scope of this check. Such error will be detected in
+          // _debugVerifyIllFatedPopulation or
+          // _debugElementsThatWillNeedToBeRebuiltDueToGlobalKeyShenanigans.
+          if (keyToParent.containsKey(key) && keyToParent[key] != parent) {
+            // We have duplication reservations for the same global key.
+            final Element older = keyToParent[key];
+            final Element newer = parent;
+            FlutterError error;
+            if (older.toString() != newer.toString()) {
+              error = FlutterError.fromParts(<DiagnosticsNode>[
+                ErrorSummary('Multiple widgets used the same GlobalKey.'),
+                ErrorDescription(
+                  'The key $key was used by multiple widgets. The parents of those widgets were:\n'
+                    '- ${older.toString()}\n'
+                    '- ${newer.toString()}\n'
+                    'A GlobalKey can only be specified on one widget at a time in the widget tree.'
+                ),
+              ]);
+            } else {
+              error = FlutterError.fromParts(<DiagnosticsNode>[
+                ErrorSummary('Multiple widgets used the same GlobalKey.'),
+                ErrorDescription(
+                  'The key $key was used by multiple widgets. The parents of those widgets were '
+                    'different widgets that both had the following description:\n'
+                    '  ${parent.toString()}\n'
+                    'A GlobalKey can only be specified on one widget at a time in the widget tree.'
+                ),
+              ]);
+            }
+            // A child can attach to two parents due to global key duplication
```

Comment:
// Fix the tree by removing the duplicated child from one of its parents to resolve the duplicated key issue. This allows us to tear down the tree during testing without producing additional misleading exceptions.